### ISSUE TITLE
Per Pixel Transparency / DisplayServer.window_set_mode sets the background to transparent. 

### ISSUE THREAD

**Issue by Alex2782** [godotengine/godot#83131]
### Godot version

v4.2.dev6.official [57a6813bb]

### System information

Godot v4.2.dev6 - macOS 13.6.0 - Vulkan (Forward+) - integrated Apple M1 - Apple M1 (8 Threads)

### Issue description

Status: the demo is fully ported and working.
https://github.com/godotengine/godot-demo-projects/issues/697

**Renderer: Forward+ and Mobile.** With 'Compatibility' everything seems to be ok.
I have already tried other 2D and 3D demos in the last weeks, such graphical problems did not occur there.

Note: when I use the system buttons to maximize the window (fullscreen), the background does not become transparent.
![image](https://github.com/godotengine/godot/assets/41921395/67e47370-65fe-4d64-acbb-7915acbd405f)

https://github.com/godotengine/godot/assets/41921395/3ceee8a9-8583-4bf9-8bc6-40d75bdf2e88

Debug: `get_viewport().transparent_bg = false`

```gdscript
func _on_Button_Fullscreen_pressed():

	print(get_viewport().transparent_bg)
	get_viewport().transparent_bg = false

	if DisplayServer.window_get_mode() == DisplayServer.WINDOW_MODE_FULLSCREEN:
		DisplayServer.window_set_mode(DisplayServer.WINDOW_MODE_WINDOWED)
	else:
		DisplayServer.window_set_mode(DisplayServer.WINDOW_MODE_FULLSCREEN)

	print(get_viewport().transparent_bg)

```
**output:**
```
false
false
```

### Steps to reproduce

1. open and start: [window_management](https://github.com/godotengine/godot-demo-projects/tree/master/misc/window_management) demo
2. push the `Fullscreen` button
3. push the `Fullscreen` button again

### Minimal reproduction project

[window_management](https://github.com/godotengine/godot-demo-projects/tree/master/misc/window_management) demo

**Comment by Calinou** [comment]
I can reproduce this in 4.2.dev4 and later (but not in 4.2.dev3 and earlier) on Linux.

**Comment by Alex2782** [comment]
**Per Pixel Transparency**

if disabled, there are no more problems with 'Fullscreen', but then the background is black when 'Transparent' `get_viewport().transparent_bg` is enabled.

![image](https://github.com/godotengine/godot/assets/41921395/b2ff102d-5e4b-4e89-9ec6-843da3a33ce8)

![Bildschirmfoto 2023-10-11 um 02 36 47](https://github.com/godotengine/godot/assets/41921395/0f1da0a4-4288-428f-b900-a13dba2d0247)

**Comment by akien-mga** [comment]
CC @bruvzg

**Comment by akien-mga** [comment]
I can't reproduce the issue on my Linux system, though Per Pixel Transparency doesn't seem to work at all for me, neither in 4.1.2 nor 4.2-beta4.

I do see a visual difference in this demo between 4.2-beta4 (top left) and 4.1.2 (bottom right):

![image](https://github.com/godotengine/godot/assets/4701338/bc233049-0d56-4edd-984c-e5f795d828ee)

**Comment by Alex2782** [comment]
**Per Pixel Transparency** active or deactivated? @akien-mga 
if I **disable** it, then it looks like the bottom right, light gray background and no graphic errors. (v4.2.beta2)

![Bildschirmfoto 2023-10-31 um 23 15 39](https://github.com/godotengine/godot/assets/41921395/64305b7b-4a61-4c46-a9a1-43d0ae28375c)


if **"Per Pixel Transparency"** is disabled and also **"Transparent"**, then the background is black.

![image](https://github.com/godotengine/godot/assets/41921395/22e65c05-5f17-48d5-94d6-9b484a9733a4)

**Comment by akien-mga** [comment]
It was activated, I tested both the window management demo, and a new project with this setting + Transparent enabled.
For me the background is always black when transparent is enabled. This is likely a separate issue, I'll open a dedicated one to track this.

**Comment by Calinou** [comment]
> I can't reproduce the issue on my Linux system, though Per Pixel Transparency doesn't seem to work at all for me, neither in 4.1.2 nor 4.2-beta4.

Is compositing enabled in your window manager? Window transparency requires it to be enabled.

**Comment by akien-mga** [comment]
> Is compositing enabled in your window manager? Window transparency requires it to be enabled (try Alt + Shift + F12 in KWin).

Good call, that was the issue. After enabling compositing with Alt+Shift+F12, I can reproduce the issue.

For me the bug doesn't even require enabling fullscreen, the demo is wrongly transparent out of the box, in a semi-transparent manner.

For each version tested, I show screenshots of the windowed mode with transparency off, then on.

In 4.1.3-stable:

![image](https://github.com/godotengine/godot/assets/4701338/dec8f163-7145-4d76-8611-577caf1ff806)
![image](https://github.com/godotengine/godot/assets/4701338/dd7abedd-77c8-4cc7-b225-4a6b701e8033)

In 4.2-dev1 and dev2:

![image](https://github.com/godotengine/godot/assets/4701338/e404595f-6f99-43bd-9b01-42683c88f23d)
![image](https://github.com/godotengine/godot/assets/4701338/235b6597-d86d-4b34-9afc-f4caa4d15b99)

In 4.2-dev3, rendering seems broken:

![image](https://github.com/godotengine/godot/assets/4701338/58aa81a5-d0b7-41fa-bd8a-7bb0345bd6ed)
![image](https://github.com/godotengine/godot/assets/4701338/dff429cd-ed8b-42e1-a16e-de4d53a6d49b)

4.2-dev4 and later:

![image](https://github.com/godotengine/godot/assets/4701338/b56df29a-11f2-4092-b2df-a66d47a5c024)
![image](https://github.com/godotengine/godot/assets/4701338/91d0cb71-822c-43cb-82ed-5de091e692b5)

**Comment by akien-mga** [comment]
> I can reproduce this in 4.2.dev4 and later (but not in 4.2.dev2 and earlier) on Linux.
> 
> I haven't managed to bisect this as the project fails to render in 4.2.dev3 (black screen for 3D rendering). A [minimal project](https://github.com/godotengine/godot/files/12863027/test_fullscreen_transparent.zip) without 3D rendering doesn't exhibit the issue.
> 
> My bisect left me on [715ebcc](https://github.com/godotengine/godot/commit/715ebcc3f15301e6fef85cdecf5acd3cd10dac55) which I'm not sure is the cause.

So I had a look too, 715ebcc is the first commit that exhibits the broken transparency because it's the first commit that fixes rendering. Before it, the project fails to render as you point out.

So it hides the real culprit. #80502 was a fixup to #80311, so I tried a local revert of both PRs, and that solves the black screen rendering, but still exhibits the broken transparency:

![image](https://github.com/godotengine/godot/assets/4701338/e9c85935-3dac-4adb-a4af-0b8c6ef9887a)

That's with current `master` plus the above two PRs reverted in reverse chronological order.

So that means they're not related to the bug, the real culprit must be another commit between 4.2.dev2 and 715ebcc.

To bisect further, you'd have to apply 715ebcc manually whenever building a commit that includes #80311 but not #80502.

**Comment by bruvzg** [comment]
> For me the bug doesn't even require enabling fullscreen, the demo is wrongly transparent out of the box, in a semi-transparent manner.

I see what's wrong with macOS code, but not sure about Linux. I think there's no way to switch transparency for the window except at window creation time on X11. If `transparent_bg` is not set, it's something wrong with the renderer.

**Comment by akien-mga** [comment]
Found the culprit in the interactive changelog for dev3, reverting #79876 fixes the issue @Calinou and I described for Linux. It might be a different problem on macOS.

**Comment by clayjohn** [comment]
https://github.com/godotengine/godot/pull/79876 just exposes the problem. Before it `transparent_bg` was totally broken and did nothing at all. 

It seems like the problem is further up the pipeline. I.e. for some reason per-pixel transparency is force enabling transparent_bg in the viewport. 

What is the intended user experience for per-pixel window transparency?

**Comment by akien-mga** [comment]
Worth noting that #80933 is also involved, to revert #79876 on top of current `master` you also need to revert #80933 first.

If I revert only #79876 and keep the inverse luminance multiplier from #80933, I still get a bug:

![image](https://github.com/godotengine/godot/assets/4701338/d24b2914-30ce-44cd-87c6-e9f9f357dfa6)

i.e. with this diff:

<details>

```diff
commit a72ca646a918951bdb81e4c4325e7b9f8caf60b8
Author: Rémi Verschelde <rverschelde@gmail.com>
Date:   Thu Nov 9 22:19:00 2023 +0100

    Revert "Fix transparent viewport backgrounds with custom clear color"
    
    This reverts commit 6effd3cde7a481b57226cf5d03c97aa5728ff7e7.

diff --git a/servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp b/servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp
index 462fc4b524..71410bd752 100644
--- a/servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp
+++ b/servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp
@@ -931,19 +931,13 @@ void RenderForwardMobile::_render_scene(RenderDataRD *p_render_data, const Color
 		{
 			// regular forward for now
 			Vector<Color> c;
-			{
-				Color cc = clear_color.srgb_to_linear() * inverse_luminance_multiplier;
-				if (rb_data.is_valid()) {
-					cc.a = 0; // For transparent viewport backgrounds.
+			c.push_back(clear_color.srgb_to_linear() * inverse_luminance_multiplier); // our render buffer
+			if (rb_data.is_valid()) {
+				if (p_render_data->render_buffers->get_msaa_3d() != RS::VIEWPORT_MSAA_DISABLED) {
+					c.push_back(clear_color.srgb_to_linear() * inverse_luminance_multiplier); // our resolve buffer
 				}
-				c.push_back(cc); // Our render buffer.
-				if (rb_data.is_valid()) {
-					if (p_render_data->render_buffers->get_msaa_3d() != RS::VIEWPORT_MSAA_DISABLED) {
-						c.push_back(clear_color.srgb_to_linear() * inverse_luminance_multiplier); // Our resolve buffer.
-					}
-					if (using_subpass_post_process) {
-						c.push_back(Color()); // Our 2D buffer we're copying into.
-					}
+				if (using_subpass_post_process) {
+					c.push_back(Color()); // our 2D buffer we're copying into
 				}
 			}
 
```

</details>

**Comment by clayjohn** [comment]
I think I know what is going on on the rendering side now. It seems to be a duplicate of https://github.com/godotengine/godot/issues/61969 Which is caused by https://github.com/godotengine/godot/pull/61109. We write to alpha even when using opaque shaders, but alpha should be ignored when we copy the 3D buffer to the 2D buffer unless using transparent background. 

In 3.x we solved this with an ``IGNORE_ALPHA`` flag in the post process shader https://github.com/godotengine/godot/pull/61382

**Comment by clayjohn** [comment]
Submitted a fix for the rendering side. Indeed the RD renderers were totally broken in multiple ways. https://github.com/godotengine/godot/pull/79876 made transparent background work, but it wasn't a full solution and wasn't fine grained enough.

The fix needs some testing in more complex projects, but works in the MRP here and in the MRP from https://github.com/godotengine/godot/issues/61969

**Comment by akien-mga** [comment]
The macOS part (original bug report) is fixed by https://github.com/godotengine/godot/pull/84683, so we can close this.

The Linux regression (not sure why it wouldn't affect other platforms, maybe it does) will be fixed by https://github.com/godotengine/godot/pull/84684, but it's a bit late to finalize it for 4.2 so that will likely be merged for 4.3 and cherry-picked for a 4.2.x release.
